polygeist/lib/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.cpp

```diff
@@ -0,0 +1,403 @@
+               << "\n";
+  os.indent(4) << "Range: ";
+  if (info.hasConstantRange()) {
+    os << "range{";
+    llvm::interleaveComma(info.getConstantRange(), os);
+    os << "}\n";
+  } else {
+    os << "<unknown>\n";
+  }
+
+  os.indent(4) << "Needs offset: " << ((info.needsOffset()) ? "Yes" : "No")
+               << "\n";
+  os.indent(4) << "Offset: ";
+  if (info.hasConstantOffset()) {
+    os << "id{";
+    llvm::interleaveComma(info.getConstantOffset(), os);
+    os << "}\n";
+  } else {
+    os << "<unknown>\n";
+  }
+
+  os.indent(4) << "Buffer: ";
+  if (info.hasKnownBuffer())
+    os << info.buffer;
+  else
+    os << "<unknown>";
+  os << "\n";
+
+  os.indent(4) << "Buffer information: ";
+  if (info.hasBufferInformation())
+    os << info.getBufferInfo();
+  else
+    os << "<unknown>";
+  os << "\n";
+
+  return os;
+}
+
+const AccessorInformation
+AccessorInformation::join(const AccessorInformation &other,
+                          AliasAnalysis &aliasAnalysis) const {
+  Value jointBuf = nullptr;
+  if (hasKnownBuffer() && other.hasKnownBuffer() &&
+      (getBuffer() == other.getBuffer() ||
+       aliasAnalysis.alias(getBuffer(), other.getBuffer()) ==
+           AliasResult::MustAlias))
+    jointBuf = getBuffer();
+
+  std::optional<BufferInformation> jointBufInfo = std::nullopt;
+  if (hasBufferInformation() && other.hasBufferInformation())
+    jointBufInfo = getBufferInfo().join(other.getBufferInfo(), aliasAnalysis);
+
+  // The range can only be omitted if this and the incoming info agree it is
+  // not needed, otherwise the conservative default is to assume the range is
+  // needed.
+  bool jointNeedsRange =
+      (needsRange() == other.needsRange()) ? needsRange() : true;
+  SmallVector<size_t, 3> jointRange = (constantRange == other.constantRange)
+                                          ? constantRange
+                                          : SmallVector<size_t, 3>{};
+
+  // The offset can only be omitted if this and the incoming info agree it is
+  // not needed, otherwise the conservative default is to assume the offset is
+  // needed.
+  bool jointNeedsOffset =
+      (needsOffset() == other.needsOffset()) ? needsOffset() : true;
+  SmallVector<size_t, 3> jointOffset = (constantOffset == other.constantOffset)
+                                           ? constantOffset
+                                           : SmallVector<size_t, 3>{};
+
+  return AccessorInformation(jointBuf, jointBufInfo, jointNeedsRange,
+                             jointRange, jointNeedsOffset, jointOffset);
+}
+
+AliasResult AccessorInformation::alias(const AccessorInformation &other,
+                                       AliasAnalysis &aliasAnalysis) const {
+  // If we can't analyze the underlying buffer, can't determine aliasing, so
+  // must assume they may alias.
+  if (!hasKnownBuffer() || !other.hasKnownBuffer())
+    return AliasResult::MayAlias;
+
+  auto isSameValue = [&](Value one, Value two) -> bool {
```

Comment:
### 
```suggestion
  bool isSameValue = [&](Value one, Value two) {
```

Reply:
type of `isSameValue` is a function type.

Reply:
Yes, `isSameValue` is a lambda function, therefore I was explicit about its return type, but used the more concise `auto` for the type in the declaration.

Reply:
sorry missed that, of course !